File: go/pkg/storage/object_store.go
// Package storage implements the ArxDB storage engine on Pebble.
//
// ObjectStore is content-addressed put/get over a sharded filesystem, matching
// the Python reference (src/arxdb/storage/object_store.py) byte-for-byte in
// its on-disk layout: blobs live under objects/xx/... where xx is the first
// two hex chars of the hash (2-char shard). Writes are atomic via temp-file +
// os.Rename.
package storage

import (
	"os"
	"path/filepath"

	"github.com/larklaflamme/arxdb/go/pkg/hashing"
)

// ObjectStore is a content-addressed, sharded filesystem blob store.
type ObjectStore struct {
	root string
}

// NewObjectStore opens (creating if needed) an object store rooted at root.
func NewObjectStore(root string) (*ObjectStore, error) {
	if err := os.MkdirAll(root, 0o755); err != nil {
		return nil, err
	}
	return &ObjectStore{root: root}, nil
}

// shardPath returns the on-disk path for a hash: objects/xx/<full-hex>.
func (o *ObjectStore) shardPath(h hashing.Hash) string {
	hexstr := h.Hex()
	return filepath.Join(o.root, hexstr[:2], hexstr)
}

// Put stores data and returns its content hash. Idempotent: re-putting an
// existing blob is a no-op.
func (o *ObjectStore) Put(data []byte) (hashing.Hash, error) {
	h := hashing.HashBytes(data)
	path := o.shardPath(h)
	if _, err := os.Stat(path); err == nil {
		return h, nil // already present
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return h, err
	}
	// Atomic write: temp file + rename.
	tmp, err := os.CreateTemp(filepath.Dir(path), ".tmp-*")
	if err != nil {
		return h, err
	}
	tmpName := tmp.Name()
	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		os.Remove(tmpName)
		return h, err
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return h, err
	}
	if err := os.Rename(tmpName, path); err != nil {
		os.Remove(tmpName)
		return h, err
	}
	return h, nil
}

// Get returns the blob for h, or nil if absent.
func (o *ObjectStore) Get(h hashing.Hash) ([]byte, error) {
	data, err := os.ReadFile(o.shardPath(h))
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return data, nil
}

// Has reports whether the blob for h is present.
func (o *ObjectStore) Has(h hashing.Hash) bool {
	_, err := os.Stat(o.shardPath(h))
	return err == nil
}

File: src/arxdb/storage/object_store.py
"""ObjectStore — content-addressed put/get over a sharded filesystem.

Blobs are immutable and idempotent: `put(x)` always returns the same hash, and
re-putting an existing blob is a no-op. Files live under `objects/xx/…` where
`xx` is the first two hex chars of the hash (2-char shard). Writes are atomic
via temp-file + `os.replace`.

Public API (Phase 1):
    ObjectStore(root: Path)
        put(data: bytes) -> Hash
        put_batch(items: list[bytes]) -> list[Hash]
        get(h: Hash) -> bytes | None
        get_batch(hashes: list[Hash]) -> list[bytes | None]
        has(h: Hash) -> bool
        has_batch(hashes: list[Hash]) -> list[bool]
"""

from __future__ import annotations

import os
import uuid
from pathlib import Path

from .hashing import Hash, hash_bytes


class ObjectStore:
    """Content-addressed blob store backed by a sharded filesystem."""

    def __init__(self, root: Path) -> None:
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)

    def _path(self, h: Hash) -> Path:
        """Filesystem path for a hash: `objects/{hex[:2]}/{hex[2:]}`."""
        hexstr = h.hex()
        return self.root / hexstr[:2] / hexstr[2:]

    def put(self, data: bytes) -> Hash:
        h = hash_bytes(data)
        dest = self._path(h)
        if dest.exists():
            return h
        dest.parent.mkdir(parents=True, exist_ok=True)
        tmp = self.root / f".tmp_{uuid.uuid4().hex}"
        try:
            with open(tmp, "wb") as f:
                f.write(data)
                f.flush()
                os.fsync(f.fileno())
            os.replace(tmp, dest)
        finally:
            if tmp.exists():
                tmp.unlink()
        return h

    def put_batch(self, items: list[bytes]) -> list[Hash]:
        return [self.put(item) for item in items]

    def get(self, h: Hash) -> bytes | None:
        dest = self._path(h)
        if not dest.exists():
            return None
        return dest.read_bytes()

    def get_batch(self, hashes: list[Hash]) -> list[bytes | None]:
        return [self.get(h) for h in hashes]

    def has(self, h: Hash) -> bool:
        return self._path(h).exists()

    def has_batch(self, hashes: list[Hash]) -> list[bool]:
        return [self.has(h) for h in hashes]
